iib/workers/tasks/build.py

```diff
@@ -337,6 +342,162 @@ def _get_resolved_image(pull_spec):
     return pull_spec_resolved
 
 
+def _get_index_database(from_index, base_dir):
+    """
+    Get database file from the specified index image and save it locally.
+
+    :param str from_index: index image to get database file from.
+    :param str base_dir: base directory to which the database file should be saved.
+    :return: path to the copied database file.
+    :rtype: str
+    :raises IIBError: if any podman command fails.
+    """
+    data = skopeo_inspect(f'docker://{from_index}')
+    try:
+        db_path = data['Labels']['operators.operatorframework.io.index.database.v1']
+    except KeyError:
+        raise IIBError('Index image doesn\'t have the label specifying its database location.')
+    _copy_files_from_image(from_index, db_path, base_dir)
+    local_path = os.path.join(base_dir, os.path.basename(db_path))
+    return local_path
+
+
+def _serve_index_registry(db_path):
```

Comment:
The [cyclomatic complexity|https://en.wikipedia.org/wiki/Cyclomatic_complexity] of this function is high, 11 (give or take). Can we make use of custom exceptions and helper methods so it's easier to read?


```python
def _serve_index_registry(db_path):
  for port in range(...):
    try:
      return (port, _serve_index_registry_at_port(db_path, port))
    except AddressAlreadyInUse:
      pass

  # If it got here, no ports are available
  raise IIBError(...)

def _serve_index_registry_at_port(db_path, port):
  for _ in range(...):
    rpc_proc = subprocess.Popen(...)
    ret = rpc_proc.poll()

    if ret is not None:
       # Process terminated unexpectedly 
       if 'address already in use' in stderr:
         raise AddressAlreadyInUse(...)
       raise IIBError(...)

     if 'serving registry' in stdout:
       return rpc_proc

     # We always want to kill the process before trying again, do it unconditionally
     rpc_proc.kill()

  # If it got here, index registry has not been initialized after all retries
  raise IIBError(...)
```

Reply:
I have simplified the code to contain less nested conditions